server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1312,7 +1312,12 @@ private void endSnapshot(SnapshotsInProgress.Entry entry, Metadata metadata, @Nu
         if (entry.isClone() && entry.state() == State.FAILED) {
             logger.debug("Removing failed snapshot clone [{}] from cluster state", entry);
             if (newFinalization) {
-                removeFailedSnapshotFromClusterState(snapshot, new SnapshotException(snapshot, entry.failure()), null);
+                removeFailedSnapshotFromClusterState(
+                    snapshot,
+                    new SnapshotException(snapshot, entry.failure()),
+                    null,
+                    ShardGenerations.EMPTY
+                );
```

Comment:
I think this is a behaviour change? Previously a failed snapshot can have individual shard snapshot being successful and in turn update other snapshot's shardGen assignments. This is especially true for clones since `buildGenerations` does not even check `Metadata` for clones. Since you said 

> On the failure paths if we're not in the process of finalization then we're not going to delete any shard gen blobs, it's ok not to update the generations of any other active shard snapshots, so ShardGenerations#EMPTY is fine

This change might be intentional. If that is the case, I find the other places where we pass `shardGenerations` to `handleFinalizationFailure` are a bit inconsistent. In both places, the failure occurs before the cleanup process can run. Hence we are not deleting any shard gen blobs either. So should we also use `ShardGenerations#EMPTY`?

Reply:
I suspect it is not possible to observe the difference in behaviour here (except ofc that the bug is fixed 😁). It doesn't really matter which shard gen we use in these failure cases, both possible blobs are valid choices, since the snapshot itself doesn't get written to the root blob.

Reply:
I'd personally prefer to have this minor consistency. But I also agree it likely won't make difference in practice.

Reply:
I dug into this some more. I think we need to use the updated `shardGenerations` in the `FinalizeSnapshotContext` listener because we might have written the new root blob before failing. The other one looks like it could be either, I'm going to leave it as-is since that's the one that's had the most thorough testing. I added comments about this in 88ff699a92f.

I'm suspicious about passing the unmodified `repositoryData` into `handleFinalizationFailure` if the `FinalizeSnapshotContext` fails, because it's possible we wrote another root blob at this point so we should really throw away all our local state and start again. As things stand I suspect we will observe the new root blob some time later and incorrectly interpret this as evidence of concurrent writes, marking the repository as corrupt. But that's a question for another day I think.